enhancements/substitutes-for.md

```diff
@@ -0,0 +1,379 @@
+upgrade into the vulnerable bundle before being able to upgrade to the patched bundle.
+* The code changes need to be contained within the catalog and opm binary so that it does not require
+OCP cluster changes.  It must work across existing 4.5+ versions of OCP clusters.
+
+### Non-Goals
+
+* Fully refactoring the upgrade graph management system out of bundle metadata so that upgrade graphs
+can be modified independently of bundle modification/publication.  (This is a future goal for OLM, however)
+
+
+## Proposal
+
+The proposal is to add a new field (tentatively named "substitutes-for") to CSVs, the value of which is one or
+more csv names that this bundle can be plugged in in place of.  This will allow existing CSVs that indicate
+they "replace" or "skip" some other CSV, to also replace/skip the new CSV that substitutes for the one
+they skip/replace.  The net result of this is the ability to publish new CSVs that can still be upgraded
+to the latest CSV, without having to update the latest CSV to add new replaces/skips references.
+
+Because the set of substitutions may grow over time, and are not guaranteed to be added in any particular 
+order, the substitutions must be applied in an order independent way.  This means each time a new 
+substitution rule is added, additional substitution rules must be generated to account for transitive
+substitutions.
+
+For example, suppose we have the following set of relationships we need to add to the catalog
+(In this example, A has a CVE, but B has already shipped as a replacement for A.  We want to patch
+A as it is still the head of some channel.  We produce A' to patch A, and then find another issue in
+A' and produce A'' to patch A'):
+
+B replaces A (original CSV relationship)
+A' substitutes for A
+A'' substitutes for A'
+
+We would first add the rule:
+A' -> A
+
+Since this is the first entry, there are no additional rules to generate.
+
+Next we add:
+A'' -> A'
+
+Which generates the following transitive substitution(D->C->B):
+A'' -> A
+
+So the total set of substitution rules is:
+A' -> A
+A'' -> A'
+A'' -> A
+
+Which means that ultimately B will be able to replace A, A', or A'', as desired.
+
+
+
+If we work the example in a different order we would first add the rule:
+A'' -> A
+
+Since this is the first entry, there are no additional rules to generate.
+
+Next we add:
+A' -> A
+
+Which generates the following transitive substitution (D->C->B):
+A'' -> A
+
+Giving us the same total set of rules:
+A' -> A
+A'' -> A'
+A'' -> A
+
+Note that the generator for the new rules is the result of a simple inner join query
+on the substitution table against itself, between the substitor and substitutee columns.
+
+Note also that each time a new rule is added, the table must be re-processed for additional
+new rules, until no new rules are generated.
+
+Alternatively the transitive substitution relationships could be derived at runtime when
```

Comment:
Do you think there are performance concerns if the index image's database gets quite large if it is computer at runtime?

Reply:
yes.  I would prefer we do this during bundle add.

That said, i don't actually expect us to maintain a huge number of bundles per operator in the catalog.  Thanks to use of skiprange, most operators only maintain one, or at most one per channel (of course for them this feature is entirely moot).

I think there will actually be a fairly small number of operators that use skips+replaces, and the number of versions of those operators that are retained in the catalog will likely be relatively small (less than a hundred).